neotron-bmc-pico/src/pwm.rs

```diff
@@ -0,0 +1,81 @@
+use stm32f0xx_hal::pac::{RCC, TIM14};
+
+#[derive(Debug, Default)]
+#[repr(align(8))]
+pub struct Config {
+	/// Whether the speaker is active
+	duration: u8,
+	/// The PWM ticks
+	ticks: u16,
+	/// The duty cycle (0 - 255)
+	duty_cycle: u8,
+}
+
+impl Config {
+    pub fn duty_cycle(&self) -> u8 {
+        self.duty_cycle
+    }
+
+    pub fn set_duty_cycle(&mut self, duty_cycle: u8) {
+        self.duty_cycle = duty_cycle;
+    }
+
+    pub fn ticks(&self) -> u16 {
+        self.ticks
+    }
+
+    pub fn set_ticks(&mut self, ticks: u16) {
+        self.ticks = ticks;
+    }
+
+    pub fn duration(&self) -> u8 {
+        self.duration
+    }
+
+    pub fn set_duration(&mut self, duration: u8) {
+        self.duration = duration;
+    }
+
+	pub fn setup_ch(&self, tim14: &TIM14) {
+		let rcc = RCC::ptr();
+		// enable and reset peripheral to a clean slate state
+		unsafe {
+			(*rcc).apb1enr.modify(|_, w| w.tim14en().set_bit());
+			(*rcc).apb1rstr.modify(|_, w| w.tim14rst().set_bit());
+			(*rcc).apb1rstr.modify(|_, w| w.tim14rst().clear_bit());
+		}
+
+		tim14
+			.ccmr1_output()
+			.modify(|_, w| w.oc1pe().set_bit().oc1m().bits(6));
+
+		// prescale 1000 (48MHz -> 48 kHz)
+		tim14.psc.write(|w| w.psc().bits(1000));
+
+		// ticks
+		tim14.arr.write(|w| unsafe { w.bits(self.ticks as u32) });
```

Comment:
Out of interesting, what `ticks` value gives you a 440 Hz tone?

Reply:
4800 / 440 ~= 109 ticks